### PR TITLE
[MNG-7957] Checksum API (backed by Resolver)

### PR DESCRIPTION
**Pull request by cstamas**
Expose checksumming API (backed by Resolver).

---

https://issues.apache.org/jira/browse/MNG-7957

### CODE REVIEW

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+     * Calculates checksums for specified file.
+     *
+     * @param file        The file for which to calculate checksums, must not be {@code null}.
+     * @param algorithms  The checksum algorithm factories to use, must not be {@code null}.
+     * @return The calculated checksums, indexed by algorithm name, or the exception that occurred while trying to
+     * calculate it, never {@code null}.
+     * @throws IOException In case of any problem.
+     */
+    @Nonnull
+    Map<String, String> calculate(@Nonnull Path file, @Nonnull Collection<ChecksumAlgorithm> algorithms)
```
Comment (gnodet) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java:85]: For ease of use ,we're missing `calculate(InputStream, Collection<ChecksumAlgorithm>)` imho.

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    /**
+     * Returns a list of {@link ChecksumAlgorithm} in same order as algorithm names are ordered, or throws if any of the
+     * algorithm name is not supported. The returned list has equal count of elements as passed in collection of names,
+     * and if names contains duplicated elements, the returned list of algorithms will have duplicates as well.
+     *
+     * @throws IllegalArgumentException if any asked algorithm name is not supported.
+     * @throws NullPointerException if passed in list of names is {@code null}.
+     */
+    @Nonnull
+    Collection<ChecksumAlgorithm> selectList(@Nonnull Collection<String> algorithmNames);
```
Comment (gnodet) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java:54]: the doc and name should be adjusted to the list -> collection change in the return type

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+     * Calculates checksums for specified data.
+     *
+     * @param data        The content for which to calculate checksums, must not be {@code null}.
+     * @param algorithms  The checksum algorithm factories to use, must not be {@code null}.
+     * @return The calculated checksums, indexed by algorithm name, or the exception that occurred while trying to
+     * calculate it, never {@code null}.
+     * @throws IOException In case of any problem.
+     */
+    @Nonnull
+    Map<String, String> calculate(@Nonnull byte[] data, @Nonnull Collection<ChecksumAlgorithm> algorithms)
```
Comment (gnodet) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java:73]: The returned Map<String, String> is not really useful.  It should be Map<ChecksumAlgorithm, String>`

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java (hunk 1 above)
Comment (gnodet) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java:86]: The returned Map<String, String> is not really useful.  It should be Map<ChecksumAlgorithm, String>`

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+     * Returns a collection of {@link ChecksumAlgorithm} in same order as algorithm names are ordered, or throws if
+     * any of the algorithm name is not supported. The returned collection has equal count of elements as passed in
+     * collection of names, and if names contains duplicated elements, the returned list of algorithms will have
+     * duplicates as well.
+     *
+     * @throws IllegalArgumentException if any asked algorithm name is not supported.
+     * @throws NullPointerException if passed in list of names is {@code null}.
+     */
+    @Nonnull
+    Collection<ChecksumAlgorithm> selectCollection(@Nonnull Collection<String> algorithmNames);
```
Comment (gnodet) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ChecksumAlgorithmService.java:56]: Would `select` be sufficient for the method name ?

File: maven-core/src/main/java/org/apache/maven/internal/impl/DefaultChecksumAlgorithmService.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    @Inject
+    public DefaultChecksumAlgorithmService(ChecksumAlgorithmFactorySelector checksumAlgorithmFactorySelector) {
+        this.checksumAlgorithmFactorySelector =
+                nonNull(checksumAlgorithmFactorySelector, "checksumAlgorithmFactorySelector");
+    }
+
+    @Override
+    public ChecksumAlgorithm select(String algorithmName) {
+        nonNull(algorithmName, "algorithmName");
+        return new DefaultChecksumAlgorithm(checksumAlgorithmFactorySelector.select(algorithmName));
```
Comment (gnodet) [inline maven-core/src/main/java/org/apache/maven/internal/impl/DefaultChecksumAlgorithmService.java:53]: What if the algorithm is not found ?

Comment (cstamas) [reply maven-core/src/main/java/org/apache/maven/internal/impl/DefaultChecksumAlgorithmService.java:53]: checksumAlgorithmFactorySelector throws IllegalArgumentEx, as documented

**Review by gnodet — APPROVED** [review]

**Pull request merged**